ouroboros-network/src/Ouroboros/Network/Subscription/Common.hs

```diff
@@ -0,0 +1,190 @@
+    -> (Socket.Socket -> IO ())
+    -> SubscriptionTarget IO Socket.SockAddr
+    -> IO ()
+subscribeTo localPort connectionAttemptDelay valencyVar k ts = do
+     x <- newTVarM []
+     doSubscribeTo x ts
+
+  where
+    doSubscribeTo :: TVar IO [ThreadId] -> SubscriptionTarget IO Socket.SockAddr -> IO ()
+    doSubscribeTo conThreads targets = do
+        target_m <- getSubscriptionTarget targets
+        case target_m of
+            Just (target, nextTargets) -> do
+                valencyLeft <- atomically $ readTVar valencyVar
+                if valencyLeft > 0
+                    then do
+                        --printf "going to subscribe to %s\n" (show target)
+                        caid <- async $ doConnect conThreads target
+                        atomically $ modifyTVar' conThreads (\t -> asyncThreadId caid:t)
+
+                        {- The time to wait depends on available deltaQ information for the
+                        -- destination address and max/min values from RFC8305.
+                        -}
+                        let delay = case connectionAttemptDelay target of
+                                         Just d  -> min maxConnectionAttemptDelay $
+                                                        max minConnectionAttemptDelay d
+                                         Nothing -> defaultConnectionAttemptDelay
+                        threadDelay delay
+                        doSubscribeTo conThreads nextTargets
+                    else do
+                        --printf "successfully started required subscriptions\n"
+                        return ()
+            Nothing -> do
+                --printf "out of targets, waiting on active connections\n"
+
+                -- We wait on the list of active connection threads instead of using an async wait
+                -- function since some of the connections may succed and then should be left
+                -- running.
+                atomically $ do
+                    activeCons  <- readTVar conThreads
+                    unless (null activeCons) retry
+                --printf "done waiting on active connections\n"
+                return ()
+
+
+    doConnect conThreads remoteAddr =
+        bracket
+            ( do
+                sd <- Socket.socket (sockAddrFamily remoteAddr) Socket.Stream
+                        Socket.defaultProtocol
+                hasRefVar <- newTVarM False
+                return (sd, hasRefVar)
+            )
+            (\(sd,hasRefVar) -> do
+                tid <- myThreadId
+                --printf "%s: dc bracket cleaning\n" $ show tid
+                Socket.close sd
+
+                atomically $ modifyTVar' conThreads (delete tid)
+                hasRef <- atomically $ readTVar hasRefVar
+                when hasRef $
+                    atomically $ modifyTVar' valencyVar (+ 1)
+            )
+            (\(sd,hasRefVar)  -> do
+                Socket.setSocketOption sd Socket.ReuseAddr 1
```

Comment:
This line is redundant, it's done again (modulo `#if`) two lines below :)

Reply:
One is for `Socket.ReuseAddr` and the other is for `Socket.ReusePort`. The part of this function that relies on `localPort` will have to be updated. Instead of a portnumber  `subscribeTo` will change to take an IPv4 address and port, and an IPv6 address and port. They port number used may not be the same.

Reply:
You're right.